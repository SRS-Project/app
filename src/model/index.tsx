export { Card } from 'src/model/Card'
export { CardType } from 'src/model/CardType'

export enum ViewState {
  Front = 'front',
  Back = 'back',
}

export enum Grade {
  Wrong,
  Correct,
}
