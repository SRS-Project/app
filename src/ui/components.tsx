export { theme } from 'src/ui/Theme'
export { Button } from 'src/ui/Button'
