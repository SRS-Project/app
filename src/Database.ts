import RxDB, { RxCollectionCreator, QueryChangeDetector, RxJsonSchema, RxDatabase, RxCollection } from 'rxdb'
import * as pouchdbAdapterIdb from 'pouchdb-adapter-idb'
import { Card } from 'src/model/interfaces'

QueryChangeDetector.enable()
QueryChangeDetector.enableDebugging()

RxDB.plugin(require('pouchdb-adapter-idb'))

const cardSchema: RxJsonSchema = {
  title: 'card schema',
  version: 0,
  type: 'object',
  properties: {
    ID: {
      type: 'string',
      primary: true,
    },
    deckID: {
      type: 'string',
      index: true,
    },
    cardTypeID: {
      type: 'number',
    },
    fields: {
      type: 'object',
      properties: {
        front: {
          type: 'string',
        },
        back: {
          type: 'string',
        },
      },
    },
    tags: {
      type: 'array',
      items: {
        type: 'string',
      },
    },
  },
  required: ['ID', 'fields'],
}

const collections: (RxCollectionCreator)[] = [
  {
    name: 'decks',
    schema: cardSchema,
    methods: {},
    // @ts-ignore
    sync: true,
  },
]

export interface CardSchema {
  ID: number
  deckID: number
  fields: string[]
  tags: string[]
}

export interface DatabaseCollection {
  heroes: RxCollection<CardSchema>
}

let dbPromise: Promise<RxDatabase> = null

const _create = async function() {
  console.log('DatabaseService: creating database..')
  const db = await RxDB.create({ name: 'srsdb', adapter: 'idb' })
  console.log('DatabaseService: created database')
  window['db'] = db // write to window for debugging

  // show leadership in title
  db.waitForLeadership().then(() => {
    console.log('isLeader now')
    document.title = 'â™› ' + document.title
  })

  // create collections
  console.log('DatabaseService: create collections')
  await Promise.all(collections.map(colData => db.collection(colData)))

  // hooks
  console.log('DatabaseService: add test data')
  await db.collections.decks.insert({
    ID: '1',
    deckID: '1',
    fields: {
      front: 'test front',
      back: 'test back',
    },
    tags: [],
  })

  return db
}

export function get() {
  if (!dbPromise) dbPromise = _create()
  return dbPromise
}
