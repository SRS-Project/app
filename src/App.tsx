import * as React from 'react'
import { theme, Button } from 'src/ui/components'
import styled, { injectGlobal } from 'styled-components'

injectGlobal`
  body {
    margin: 0;
  }
`

const StyledApp = styled.div`
  font-family: ${theme.fonts.body};
  background-color: ${theme.colors.dark};
  width: 100vw;
  height: 100vh;
`

export class App extends React.Component<{}, {}> {
  render() {
    return (
      <StyledApp>
        <Button label={'Hey there from App!'} />
      </StyledApp>
    )
  }
}
