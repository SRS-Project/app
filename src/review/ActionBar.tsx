import * as React from 'react'
import { ViewState } from 'src/model/interfaces'
import { Button } from 'src/ui/components'

export interface ActionBarProps {
  viewState: ViewState
  revealHandler: () => void
}

export class ActionBar extends React.Component<ActionBarProps, {}> {
  renderFrontActions = () => {
    return <Button label="Reveal card" onClick={this.props.revealHandler} />
  }

  renderBackActions = () => {
    return (
      <>
        <Button label="Wrong" />
        <Button label="Correct" />
      </>
    )
  }

  render() {
    switch (this.props.viewState) {
      case ViewState.Front:
        return this.renderFrontActions()
      case ViewState.Back:
        return this.renderBackActions()
    }
  }
}
