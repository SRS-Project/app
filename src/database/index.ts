import RxDB, { RxDatabaseCreator, RxCollectionCreator, QueryChangeDetector, RxDatabase, RxCollection } from 'rxdb'
import * as pouchdbAdapterIdb from 'pouchdb-adapter-idb'

import { Card, CardType } from 'src/model'
import { CardCollectionCreator, CardTypeCollectionCreator } from 'src/database/schema'
import { seed } from 'src/database/seed'

const databaseConfig: RxDatabaseCreator = { name: 'srsdb', adapter: 'idb' }

QueryChangeDetector.enable()
QueryChangeDetector.enableDebugging()

RxDB.plugin(require('pouchdb-adapter-idb'))

const collections: (RxCollectionCreator)[] = [CardCollectionCreator, CardTypeCollectionCreator]

export interface DatabaseCollection {
  cards: RxCollection<Card>
  cardTypes: RxCollection<CardType>
}

const createDatabase = async function(creator: RxDatabaseCreator) {
  const db = await RxDB.create(creator)

  await db.waitForLeadership()
  await Promise.all(collections.map(colData => db.collection(colData)))
  await seed(db)

  return db
}

let databasePromise: Promise<RxDatabase> = null

export function get() {
  if (!databasePromise) {
    databasePromise = createDatabase(databaseConfig)
  }

  return databasePromise
}
