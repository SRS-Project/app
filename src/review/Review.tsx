import * as React from 'react'
import { Card, ViewState, Grade } from 'src/model/interfaces'
import { CardRenderer } from 'src/cards/CardRenderer'
import { ActionBar } from 'src/review/ActionBar'

export interface ReviewProps {
  card?: Card
  viewState?: ViewState
  handleReveal: () => void
  handleGrade: (grade: Grade) => void
}

export const Review: React.SFC<ReviewProps> = ({ card, viewState, handleReveal, handleGrade }) => {
  if (!card) {
    return <>Nothing to review.</>
  }

  return (
    <>
      <CardRenderer card={card} viewState={viewState} />
      <ActionBar viewState={viewState} handleReveal={handleReveal} handleGrade={handleGrade} />
    </>
  )
}

Review.defaultProps = {
  viewState: ViewState.Front,
}
