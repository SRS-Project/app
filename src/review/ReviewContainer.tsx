import * as React from 'react'
import { Card, ViewState, Grade } from 'src/model/interfaces'
import { Review } from 'src/review/Review'

export interface ReviewState {
  viewState: ViewState
  time: number
}

interface State {
  card?: Card
  reviewState?: ReviewState
}

export class ReviewContainer extends React.Component<{}, {}> {
  state: State = {
    reviewState: {
      viewState: ViewState.Front,
      time: 0,
    },
    card: {
      deckID: 1,
      cardTypeID: 1,
      fields: {
        front: 'こんにちは',
        back: 'Hello',
      },
      tags: [] as string[],
    },
  }

  reveal = () => {
    this.setState({
      ...this.state,
      reviewState: {
        ...this.state.reviewState,
        viewState: ViewState.Back,
      },
    })
  }

  grade = (grade: Grade) => {
    this.setState({
      card: null,
      reviewState: {
        viewState: ViewState.Front,
        time: 0,
      },
    })
  }

  render() {
    const { card, reviewState } = this.state

    return (
      <Review card={card} viewState={reviewState.viewState} revealHandler={this.reveal} gradeHandler={this.grade} />
    )
  }
}
