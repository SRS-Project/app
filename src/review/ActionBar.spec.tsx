import * as React from 'react'
import { create } from 'react-test-renderer'
import { shallow } from 'enzyme'

import { ActionBar } from 'src/review/ActionBar'
import { ViewState } from 'src/model'
import { Button } from 'src/ui/components'

describe('ActionBar', () => {
  const initialProps = {
    viewState: ViewState.Front,
    handleReveal: jest.fn(),
    handleGrade: jest.fn(),
  }

  test('shows `reveal state` button on init', () => {
    const component = <ActionBar {...initialProps} />

    let tree = create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('shows grading buttons on reveal', () => {
    const props = {
      ...initialProps,
      viewState: ViewState.Back,
    }
    const component = create(<ActionBar {...props} />)

    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('triggers handleReveal after clicking reveal', () => {
    const props = initialProps
    const tree = shallow(<ActionBar {...props} />)

    tree.find(Button).simulate('click')
    expect(props.handleReveal).toHaveBeenCalled()
  })

  test('triggers handleGrade when clicking grading button', () => {
    const props = {
      ...initialProps,
      viewState: ViewState.Back,
    }
    const tree = shallow(<ActionBar {...props} />)

    tree.find(Button).forEach(button => button.simulate('click'))
    expect(props.handleGrade).toHaveBeenCalledTimes(2)
  })
})
