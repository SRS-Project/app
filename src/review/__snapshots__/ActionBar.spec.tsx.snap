// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionBar shows \`reveal state\` button on init 1`] = `"reveal card"`;
